### ISSUE TITLE
Convert arb to integer

### ISSUE THREAD

**Issue by 8d1h** [Nemocas/Nemo.jl#1128]
Hi, I would expect the following to work
```julia
julia> RR = ArbField(100);

julia> fmpz(RR(1))
1

julia> fmpz(RR(1)/3*3)
ERROR: Argument must represent a unique integer
```
The code looks suspicious https://github.com/Nemocas/Nemo.jl/blob/0e2cbec1f390f0acb6d4124901c67538d9409b65/src/arb/arb.jl#L142
```julia
function fmpz(x::arb)
  !isexact(x) && error("Argument must represent a unique integer")
  (b, n) = unique_integer(x)
  !b ? error("Argument must represent a unique integer") : return n
end
```

**Comment by tthsqe12** [comment]
Everything is working here as expected: the julia convention is to throw when the conversion is nonexact.

**Comment by 8d1h** [comment]
Well first julia does not throw in the following case (although strictly speaking it's not the same thing since julia does not have error bounds)
```julia
julia> Int(1/3*3)
1
```
Also I guess the docstring is misleading
> Return x as an fmpz if it represents an unique integer, else throws an error.

It should say "if x *is* an exact integer".

Although I would still argue that `ZZ(x)` or `fmpz(x)` is a more sensible thing to do than `unique_integer(x)[2]`...

**Comment by tthsqe12** [comment]
For `Int(1/3*3)` the `1/3` is represented as a `Float64` with error but the multiplication `1/3*3` *rounds* the answer back to the exact `1` as a `Float64`. So again, everything is working as expected. Machine precision floats are not incredibly usefull for rigorous calculation:
```julia
julia> 3/7 - 3*(1/7)
0.0

julia> 5/7 - 5*(1/7)
1.1102230246251565e-16
```

As for the names, that is up to the maintainers here ...

**Comment by 8d1h** [comment]
> Machine precision floats are not incredibly usefull for rigorous calculation:

Guess you are right. I just stared at the following for a long time
```julia
julia> RR = ArbField(200);

julia> a = RR(0.1)
0.1000000000000000055511151231257827021181583404541015625000000

julia> isexact(a)
true

julia> b = RR("0.1")
[0.10000000000000000000000000000000000000000000000000000000000 +/- 1.41e-61]

julia> isexact(b)
false
```
And the answer is just the word "binary"... It'd be great if this were explained better in the docs (and with examples).

**Comment by tthsqe12** [comment]
Yeah, I have had my julia fingers burned on multiple occasions on stuff exactly like `RR(0.1)`. I think I did `x=RR(sqrt(2))` (instead of `x=sqrt(RR(2))`) and then it said `x` was exact ... how the *expletive* is `x` exact ... oohhhh ...

**Comment by thofma** [comment]
I changed the title to reflect the actual issue.

I don't see the problem with the other examples. Usual problems with floating point values.

**Comment by tthsqe12** [comment]
I guess the two things are:
- `fmpz(RR(1)/3*3)` should thow an "inexact error" with a clearer error message
- `(::ArbField)(::Float64)` creates an exact arb, which can be considered counterintuitive

**Comment by thofma** [comment]
Even if your precision is less than 53?

**Comment by tthsqe12** [comment]
Correction: can create

**Comment by 8d1h** [comment]
> I changed the title to reflect the actual issue.

Thanks!

> `(::ArbField)(::Float64)` creates an exact arb, which can be considered counterintuitive

Yes I agree on this point. When I write `RR(0.1)`, what I have in mind is perhaps `[0.10000000 +/- ε]` with some error, but certainly not the exact value `3602879701896397//36028797018963968`... (although technically this *is* what I'm writing...) I think it makes sense to introduce the error term when converting a float to `arb`.

**Comment by thofma** [comment]
0.1 is a literal which can not be distinguished from 3602879701896397//36028797018963968. This value can be represented with a ball of radius 0 if the precision is bigger than 53, which it is in your examples. There is no error term.

**Comment by tthsqe12** [comment]
So, I think having RR(0.1) always return a nonexact arb is going to be a hard sell to the julia folks: they really do think of them as rationals - just try `Rational{BigInt}(0.1)`. Therefore I see the resolution of these issues as:
- have `ZZ(RR(1)/3*3)` throw an inexact error with a better message
- complete the fmpq interface by implementing `QQ(0.1)` ect

Is this acceptable?

**Comment by thofma** [comment]
Acceptable for me.
